modules/activiti-engine/src/main/java/org/activiti/engine/impl/history/DefaultHistoryManager.java

```diff
@@ -423,7 +429,11 @@ public void recordExecutionReplacedBy(ExecutionEntity execution, InterpretableEx
         .executionId(execution.getId())
         .list();
       for (HistoricVariableInstanceEntity historicVariableInstance: historicVariableInstances) {
-          historicVariableInstance.setExecutionId(replacedBy.getId());
+    	  // only variables of unfinished tasks
+    	  if (unfinishedTasks.contains(historicVariableInstance.getTaskId())) {
```

Comment:
We need to consider the cached entities that is not flushed as well, so we need to add this filter to the cached entities as well. Specifically, it is the following part.
https://github.com/Activiti/Activiti/blob/a43ff77ed55f881abeb7da87125b29f0bbd30df6/modules/activiti-engine/src/main/java/org/activiti/engine/impl/history/DefaultHistoryManager.java#L421

Reply:
I've also included cached tasks to the check